i18n(es): translate `actions.mdx` (#11442)
withastro/docs: src/content/docs/es/guides/actions.mdx

Usa acciones en lugar de puntos de conexión API para una comunicación fluida entre tu código de cliente y servidor y para:

- Validar automáticamente las entradas de datos JSON y de formularios utilizando la [validación de Zod](https://zod.dev/?id=primitives).
- Generar funciones con seguridad de tipos para llamar a tu backend desde el cliente e incluso [desde acciones de formularios HTML](#llamar-a-acciones-desde-una-acción-de-formulario-html). No necesitas llamadas manuales a `fetch()`.
- Estandarizar errores de backend con el objeto [`ActionError`](/es/reference/modules/astro-actions/#actionerror).

## Uso básico

Las acciones se definen en un objeto `server` exportado desde `src/actions/index.ts`:

```ts title="src/actions/index.ts"
import { defineAction } from 'astro:actions';
import { z } from 'astro:schema';

export const server = {
  myAction: defineAction({ /* ... */ })
}
```

> Tus acciones están disponibles como funciones desde el módulo `astro:actions`. Importa `actions` y llámalas del lado del cliente dentro de un [componente de framework de interfaz de usuario](/es/guides/framework-components/), [una solicitud POST de formulario](#llamar-a-acciones-desde-una-accion-de-formulario-html), o usando una etiqueta `<script>` en un componente Astro.

Reviewer: ```suggestion
Tus acciones están disponibles como funciones desde el módulo `astro:actions`. Importa `actions` y llámalas del lado del cliente dentro de un [componente de framework de interfaz de usuario](/es/guides/framework-components/), [una solicitud POST de formulario](#llamar-a-acciones-desde-una-acción-de-formulario-html), o usando una etiqueta `<script>` en un componente Astro.
```

En el siguiente ejemplo, un componente `LikeButton.tsx` llama a la acción `likePost()` cuando se hace clic. Si ocurre un error de autenticación, se utiliza el atributo `error.code` para determinar si se debe mostrar un enlace de inicio de sesión:

```tsx title=src/components/LikeButton.tsx ins="if (error.code === 'UNAUTHORIZED') setShowLogin(true);"
import { actions } from 'astro:actions';
import { useState } from 'preact/hooks';

export function LikeButton({ postId }: { postId: string }) {
  const [showLogin, setShowLogin] = useState(false);
  return (
    <>
      {
        showLogin && <a href="/signin">Log in to like a post.</a>
      }
      <button onClick={async () => {
        const { data, error } = await actions.likePost({ postId });
        if (error?.code === 'UNAUTHORIZED') setShowLogin(true);
        // Retorno temprano para errores inesperados
        else if (error) return;
        // actualizar "me gusta"
      }}>
        Like
      </button>
    </>
  )
}
```

### Manejo de redirecciones del cliente

> Al llamar a acciones desde el cliente, puedes integrarte con una biblioteca del lado del cliente como `react-router`, o puedes usar la función [`navigate()` de Astro](/es/guides/view-transitions/#activar-la-navegacion) para redirigir a una nueva página cuando una acción tiene éxito.

Reviewer: ```suggestion
Al llamar a acciones desde el cliente, puedes integrarte con una biblioteca del lado del cliente como `react-router`, o puedes usar la función [`navigate()` de Astro](/es/guides/view-transitions/#desencadenar-la-navegación) para redirigir a una nueva página cuando una acción tiene éxito.
```

3.  Agrega un `<script>` al formulario HTML para enviar la entrada del usuario. Este ejemplo anula el comportamiento de envío predeterminado del formulario para llamar a `actions.newsletter()` y redirige a `/confirmation` usando la función `navigate()`:

    ```astro title=src/components/Newsletter.astro ins={12-23} collapse={2-8}
    <form>
      <label for="email">E-mail</label>
      <input id="email" required type="email" name="email" />
      <label>
        <input required type="checkbox" name="terms">
        I agree to the terms of service
      </label>
      <button>Sign up</button>
    </form>

    <script>
      import { actions } from 'astro:actions';
      import { navigate } from 'astro:transitions/client';

      const form = document.querySelector('form');
      form?.addEventListener('submit', async (event) => {
        event.preventDefault();
        const formData = new FormData(form);
        const { error } = await actions.newsletter(formData);
        if (!error) navigate('/confirmation');
      })
    </script>
    ```

>     <ReadMore>Consulta [“Llamar a acciones desde una acción de formulario HTML”](#llamar-a-acciones-desde-una-accion-de-formulario-html) para conocer una forma alternativa de enviar datos de formulario.</ReadMore>

Reviewer: ```suggestion
    <ReadMore>Consulta [“Llamar a acciones desde una acción de formulario HTML”](#llamar-a-acciones-desde-una-acción-de-formulario-html) para conocer una forma alternativa de enviar datos de formulario.</ReadMore>
```

Para recuperar los errores de entrada, usa la utilidad `isInputError()` para verificar si un error fue causado por una entrada no válida. Los errores de entrada contienen un objeto `fields` con mensajes para cada nombre de entrada que no se pudo validar. Puedes usar estos mensajes para solicitar al usuario que corrija su envío.

El siguiente ejemplo verifica el error con `isInputError()`, luego verifica si el error está en el campo de correo electrónico, antes de finalmente crear un mensaje a partir de los errores. Puedes usar la manipulación del DOM de JavaScript o tu framework de interfaz de usuario preferido para mostrar este mensaje a los usuarios.

```js /isInputError(?= )/ {5-12}
import { actions, isInputError } from 'astro:actions';

const form = document.querySelector('form');
const formData = new FormData(form);
const { error } = await actions.newsletter(formData);
if (isInputError(error)) {
  // Manejar errores de entrada.
  if (error.fields.email) {
    const message = error.fields.email.join(', ');
  }
}
```

## Llamar a acciones desde una acción de formulario HTML

:::note

> Las páginas deben renderizarse bajo demanda al llamar a acciones usando una acción de formulario. [Asegúrate de que la pre-renderización esté deshabilitada en la página](/es/guides/on-demand-rendering/#habilitacion-del-renderizado-bajo-demanda) antes de usar esta API.

Reviewer: ```suggestion
Las páginas deben renderizarse bajo demanda al llamar a acciones usando una acción de formulario. [Asegúrate de que la pre-renderización esté deshabilitada en la página](/es/guides/on-demand-rendering/#habilitar-el-renderizado-bajo-demanda) antes de usar esta API.
```

Usa la propiedad `data` para construir una URL para usar con `Astro.redirect()`:

```astro title="src/pages/products/create.astro" {4-7}
---
import { actions } from 'astro:actions';

const result = Astro.getActionResult(actions.createProduct);
if (result && !result.error) {
  return Astro.redirect(`/products/${result.data.id}`);
}
---

<form method="POST" action={actions.createProduct}>
  </form>
```

### Manejar errores de acción de formulario

Llamar a `Astro.getActionResult()` en el componente Astro que contiene tu formulario te da acceso a los objetos `data` y `error` para el manejo de errores personalizado.

El siguiente ejemplo muestra un mensaje de falla general cuando falla una acción `newsletter`:

```astro title="src/pages/index.astro" {4,7-9}
---
import { actions } from 'astro:actions';

const result = Astro.getActionResult(actions.newsletter);
---

{result?.error && (
  <p class="error">Unable to sign up. Please try again later.</p>
)}
<form method="POST" action={actions.newsletter}>
  <label>
    E-mail
    <input required type="email" name="email" />
  </label>
  <button>Sign up</button>
</form>
```

> Para una mayor personalización, puedes [usar la utilidad `isInputError()`](#visualizacion-de-errores-de-entrada-de-formulario) para verificar si un error es causado por una entrada no válida.

Reviewer: ```suggestion
Para una mayor personalización, puedes [usar la utilidad `isInputError()`](#visualización-de-errores-de-entrada-de-formulario) para verificar si un error es causado por una entrada no válida.
```

El siguiente ejemplo representa un banner de error debajo del campo de entrada `email` cuando se envía un correo electrónico no válido:

```astro title="src/pages/index.astro" ins={5,13} ins='aria-describedby="error"'
---
import { actions, isInputError } from 'astro:actions';

const result = Astro.getActionResult(actions.newsletter);
const inputErrors = isInputError(result?.error) ? result.error.fields : {};
---

<form method="POST" action={actions.newsletter}>
  <label>
    E-mail
    <input required type="email" name="email" aria-describedby="error" />
  </label>
  {inputErrors.email && <p id="error">{inputErrors.email.join(',')}</p>}
  <button>Sign up</button>
</form>
```

#### Conservar valores de entrada en caso de error

> Las entradas se borrarán cada vez que se envíe un formulario. Para conservar los valores de entrada, puedes [habilitar transiciones de vista](/es/guides/view-transitions/#agregar-transiciones-de-vista-a-una-pagina) en la página y aplicar la directiva `transition:persist` a cada entrada:

Reviewer: ```suggestion
Las entradas se borrarán cada vez que se envíe un formulario. Para conservar los valores de entrada, puedes [habilitar transiciones de vista](/es/guides/view-transitions/#agregando-las-view-transitions-a-una-página) en la página y aplicar la directiva `transition:persist` a cada entrada:
```

El siguiente ejemplo rechaza todas las solicitudes de acciones que no tienen un token de sesión válido. Si falla la verificación, se devuelve una respuesta "Prohibido". Nota: este método garantiza que las acciones solo sean accesibles cuando hay una sesión presente, pero **no** sustituye la autorización segura.

```ts title="src/middleware.ts"
import { defineMiddleware } from 'astro:middleware';
import { getActionContext } from 'astro:actions';

export const onRequest = defineMiddleware(async (context, next) => {
  const { action } = getActionContext(context);
  // Verifica si se llamó a la acción desde una función del lado del cliente
  if (action?.calledFrom === 'rpc') {
    // Si es así, busca un token de sesión de usuario
    if (!context.cookies.has('user-session')) {
      return new Response('Forbidden', { status: 403 });
    }
  }

  context.cookies.set('user-session', /* token de sesión */);
  return next();
});
```

## Llamar a acciones desde componentes Astro y puntos de conexión del servidor

> Puedes llamar a acciones directamente desde los scripts de componentes Astro usando el contenedor `Astro.callAction()` (o `context.callAction()` cuando usas un [punto de conexión del servidor](/es/guides/endpoints/#puntos-de-conexion-de-servidor-rutas-api)). Esto es común para reutilizar la lógica de tus acciones en otro código de servidor.

Reviewer: ```suggestion
Puedes llamar a acciones directamente desde los scripts de componentes Astro usando el contenedor `Astro.callAction()` (o `context.callAction()` cuando usas un [punto de conexión del servidor](/es/guides/endpoints/#endpoints-del-servidor-rutas-de-api)). Esto es común para reutilizar la lógica de tus acciones en otro código de servidor.
```

> Puedes llamar a acciones directamente desde los scripts de componentes Astro usando el contenedor `Astro.callAction()` (o `context.callAction()` cuando usas un [punto de conexión del servidor](/es/guides/endpoints/#puntos-de-conexion-de-servidor-rutas-api)). Esto es común para reutilizar la lógica de tus acciones en otro código de servidor.

Reviewer: Generally translations, and the other translated Spanish pages don't translate "endpoint" and keep it as is, I would appreciate if you could update this page to reflect that so we keep it consistent.
